client/connection/manager.go

```diff
@@ -55,33 +68,72 @@ func (manager *connectionManager) Connect(consumerID, providerID identity.Identi
 		}
 	}()
 
-	proposal, err := manager.findProposalByProviderID(providerID)
+	cancelable := utils.NewCancelable()
+	manager.closeAction = utils.CallOnce(func() {
+		log.Info(managerLogPrefix, "Closing active connection")
+		manager.status = statusDisconnecting()
+		cancelable.Cancel()
+	})
+
+	val, err := cancelable.
+		Request(func() (interface{}, error) {
+			return manager.findProposalByProviderID(providerID)
+		}).
+		Call()
 	if err != nil {
 		return err
 	}
+	proposal := val.(*dto.ServiceProposal)
 
-	dialogEstablisher := manager.newDialogEstablisher(consumerID)
-	manager.dialog, err = dialogEstablisher.EstablishDialog(providerID, proposal.ProviderContacts[0])
+	val, err = cancelable.
+		Request(func() (interface{}, error) {
+			return manager.newDialog(consumerID, providerID, proposal.ProviderContacts[0])
+		}).
+		Cleanup(utils.SkipOnError(func(val interface{}) {
```

Comment:
This error skipping looks complicated:
```
Cleanup(utils.SkipOnError(...))
```

`cancelable.Done()` - executes on success only
`cancelable.Always()` - executes on errors also

Reply:
Now it's Cleanup(utils.InvokeOnSuccess(...)) because it's not a part of cancelable request, but a function decorator which invoces callback only if error == nil

Reply:
I know pattern is applied, but still complicated

Reply:
either if err == nil then do cleanup or function decorator. I would stay with decorator